### DISCUSSION TITLE
v23无法自动挂载非系统分区

### DISCUSSION THREAD

**Discussion by fslong520** [linuxdeepin/developer-center#3974] [General | 常规讨论]
目前是双系统，三块硬盘。
在v23下开机只能自动挂载v23的分区，其余分区均不能自动挂载。
文件管理器中的自动挂载设置项设置了也无效，反倒是移动硬盘可以自动挂载，内置硬盘不能自动挂载。

20.8系统如果boot分区是单独挂载的，是可以自动挂载其他分区的，包括Windows的，但到了23这样也不行。
如果是一根到底的分区方式，20.8也是无法挂载其他分区。

用修改fstab文件的方式20.8和23都是可以自动挂载的，但倾向于不使用这种方式，这是比较敏感的操作，普通用户也不建议改这个。

请问：
1. 为何20.8单独挂载boot分区就可以自动挂载其他分区了，不单独挂载boot分区就不行？单独挂载boot分区跟一根到底的分区方式启动时候做的事情有什么不同吗？
2. 这是什么原因导致v23即使单独挂载boot分区也不能自动挂载的？
3. 如果我不想修改fstab文件还想自动挂载其他分区，需要做什么？

**Comment by BLumia** [comment]
1 和 2 我不太清楚，但 1 所描述的行为本身很奇怪。单独挂载 boot 不应当同时自动挂载其它分区（i.e. 可能是 bug，如果可复现的话可以按 bug 反馈）。

3 的话，有两个策略。一个是文管应该是本身自带一个自动挂载的选项的，可以直接用它。另一个是用分区管理工具（我不太确定目前深度的分区管理工具有没有这个功能，如果没有的话可以看看别的），对于分区应该有个自动挂载的选项，通过它设置。

**Reply by fslong520** [reply]
你能不能认真看一下？
1、1中说的是安装系统的时候boot单独挂载一个分区，而不是只分一个根这种分区方式装系统。在20.8下，只要boot分区是单独的，就可以自动挂载，其他分区。
2、我都说了系统文管的自动挂载没用才来这问，deepin的磁盘管理工具有没有这功能你不知道？打开看一眼就知道没有呀。其他分区工具也得开机启动之后点开才行呀。其他分区工具打开时候就需要鉴权呀，就是这个挂载内置分区还需要鉴权比较反人类。

**Reply by BLumia** [reply]
抱歉，最近处理的 issue 繁多，看的粗略了点。但重新审视此主题后仍然觉得有问题。

> 在20.8下，只要boot分区是单独的，就可以自动挂载，其他分区。

但这个描述和你的最初描述一样，仍然很奇怪。自动挂载的行为本身不应当和 boot 是否是单独的分区相关。如果确实行为如你描述所述，那应当是 bug。

> 系统文管的自动挂载没用才来这问

这需要看具体的日志才能判断为什么没有用。我曾经维护文件管理器项目（v15）的时候，自动挂载的行为本身是文件管理器的 dock 插件在 dock 加载对应插件时进行的，如果这个行为没有变化的话，更不可能因为 boot 分区是否单独划分而有行为差异。如果文管项目后续有调整相关行为，可能需要另外具体判断原因，但无论如何，如果你描述属实，那极有可能是 bug。我还是建议作为 bug 开 issue 进行反馈（自然，开 bug 的时候就会需要提供一些具体的信息了，包括详细的复现步骤和日志之类）。

> deepin的磁盘管理工具有没有这功能你不知道？打开看一眼就知道没有呀。

只是因为我的开发环境甚至在用 commit 仓库，和实际已发布情况可能不符，所以没以我的环境为准。

> 其他分区工具也得开机启动之后点开才行呀。其他分区工具打开时候就需要鉴权呀，就是这个挂载内置分区还需要鉴权比较反人类。

如果没记错的话（注：记忆也来自 v15 维护文管时），实际上，udisks2 提供了标记一个分区是否应当自动挂载的功能（`HintAuto`），这个行为最终会由 udisks 对应的服务去进行，而不是分区工具。分区工具实质上只是提供了配置这个选项的一个界面功能而已。v15 当时我甚至有规划将文管的自动挂载调整至此设置项，但后续这个项目交手后就没再进行相关的工作了。

最近 v23 beta 版本临近，要处理的事情比较多而杂，造成这样的回答可能仍然有些仓促，如果有任何不恰当的回答还请指出，感激不尽！

**Reply by BLumia** [reply]
关于自动挂载失效的问题，也建议 @babyfengfjx 关注一下，看看是否能根据已有的描述复现问题。

**Reply by fslong520** [reply]
抱歉，下午我可能也有点言辞激烈，因为这问题其实困扰挺久了，如果能解决就不上这里麻烦大家了，所以当时也有些冲动，再说一次抱歉。

这个问题其实挺久了，论坛发帖后也有别的同学说会这种问题，我这边是必先的问题。
大致的来龙去脉如下：
1. 我一般安装系统的时候不挂载到数据盘到`/data`，因为有时候会在其他系统操作。分区方案一般是`efi`、`/`或者`efi`、`/`、`/home`。
2. 这种分区方案大概在20.5还是20.6时期文管的自动挂载就已经失效。需要开机打开文管，双击那个磁盘才会挂载，但不需要鉴权。
3. 当时我还在用better-dde，以为是他的问题，并未提交bug。
4. 到了20.8的时候不但自动挂载失效，并且手动挂载时候还需要鉴权，某次我安装系统的时候分区方案使用的是`efi`、`/boot、`/`，意外发现这种分区方案下是可以自动挂载的，猜测全盘安装方式自动挂载也是正常的。

经过晚上测试，`openkylin`、`deepin v23、`ubuntu`同样的也有此问题，自动挂载会失效，手动挂载同样也需要鉴权。
`manjaro`和`mint`虽然也无法自动挂载，但在挂载内置磁盘的时候不需要鉴权，跟上面叙述的deepin在20.5时期状态差不多，对比发现：
1. `manjaro`的磁盘是挂载在`/run/media`下面的，而deepin是在`/media`下面。
2. `manjaro`在`/etc/polkit-1/rules.d`会多一条名为`99-manjaro.rules`的规则，这文件内容较多，跟`udisks2`有关的内容是：
```js
polkit.addRule(function(action, subject) {
    if (action.id.indexOf("org.freedesktop.udisks2.") == 0 && subject.isInGroup("wheel")) {
        return polkit.Result.YES;
    }
});
```
3. 给deepin新增这条规则依然是挂载内置磁盘需要鉴权，并无作用。查阅有关资料了解到似乎是与`polkit`有关，再联想到在非dde桌面环境下，比如deepin v23切换到`kde`、`mate`、`cinnamon`之后鉴权均会失效无法完成鉴权，个人猜测这里应该是`polkit`有问题，希望您这边进一步排查，到这我就进行不下去了。

**Reply by tsic404** [reply]
```
polkit.addRule(function(action, subject) {
    if (action.id.indexOf("org.freedesktop.udisks2.") == 0 && subject.isInGroup("sudo")) {
        return polkit.Result.YES;
    }
});
```

用户组的问题吧，deepin 23 没有 wheel 组，是 sudo 组。

**Reply by fslong520** [reply]
感谢，我去测试下。

**Comment by babyfengfjx** [comment]
确认这个现象在V20版本不存在，只有V23版本有该问题，上周讨论来看社区成员应该是不知道该变化的来由，目前是提单到文管作进一步分析，我看世龙已经提单到内测，我们已经跟进，后续有相关信息会作相应同步。

**Comment by itsXuSt** [comment]
V23 仓库中的 udisks2 没有应用这个补丁导致的。

```
--- udisks2-2.8.1.orig/data/org.freedesktop.UDisks2.policy.in
+++ udisks2-2.8.1/data/org.freedesktop.UDisks2.policy.in
@@ -29,7 +29,7 @@
     <defaults>
       <allow_any>auth_admin</allow_any>
       <allow_inactive>auth_admin</allow_inactive>
-      <allow_active>auth_admin_keep</allow_active>
+      <allow_active>yes</allow_active>
     </defaults>
   </action>
```

**Reply by BLumia** [reply]
@Zeno-sole 我看了下 debian 的 `libudisks2-0 (2.9.4-4)` 也不存在这个 patch。我们应该打这个 patch 吗？

感觉还是不太对，这里是打到 `<action id="org.freedesktop.udisks2.filesystem-mount-system">` 上的，但这里也许不应当被视为系统卷？v20 还有打别的 patch 吗？

**Reply by Zeno-sole** [reply]
v20 的udisks2 魔改很多。。

**Reply by fslong520** [reply]
我排查是没有写rules，权限问题几率较大。

**Reply by itsXuSt** [reply]
> @Zeno-sole 我看了下 debian 的 `libudisks2-0 (2.9.4-4)` 也不存在这个 patch。我们应该打这个 patch 吗？
> 
> 感觉还是不太对，这里是打到 `<action id="org.freedesktop.udisks2.filesystem-mount-system">` 上的，但这里也许不应当被视为系统卷？v20 还有打别的 patch 吗？

我查了 V20 apt source 下来的源码和 V23 的，比较了下，V23 是相对纯净的，和上游 storage project 项目下的 udisks 基本一致，v20 的 debian 文件夹里是包含了这个补丁的。
它问题里提的是内置磁盘，在 udisks 执行挂载的时候，传给 polkit 的就是 mount-system 这个 action_id。如果是普通的外接磁盘的话肯定就不是这个了。

**Reply by itsXuSt** [reply]
> @Zeno-sole 我看了下 debian 的 `libudisks2-0 (2.9.4-4)` 也不存在这个 patch。我们应该打这个 patch 吗？
> 
> 感觉还是不太对，这里是打到 `<action id="org.freedesktop.udisks2.filesystem-mount-system">` 上的，但这里也许不应当被视为系统卷？v20 还有打别的 patch 吗？

![image](https://user-images.githubusercontent.com/50353838/230832631-945935ec-0b57-46b6-ab2a-338d2bfbb51f.png)
除了这个补丁之外，还有两个，一个是测试相关的，另一个是移动了 DBus 文件的位置。

**Comment by SuperEffie** [comment]
【验证环境】
【验证镜像】：https://packagesupdate.uniontech.com/repo/uos-professional/image-beige/CUSTOM/lichenggang/20230309/
【集成链接】：https://github.com/orgs/linuxdeepin/projects/21/views/1?pane=issue&itemId=25203539
【版本信息】：udisks2：2.9.4.1-deepin1
【验证结论】：测试通过。PC上安装多系统，在deepin v23系统下挂载其他系统盘，可成功挂载。
【录屏信息】：
![挂载](https://user-images.githubusercontent.com/58986464/230874122-94b69d52-2b2e-4bcb-9eeb-6ed4f209667b.gif)
